opts/hostcfg.go

```diff
@@ -82,43 +85,101 @@ func (i *IPAddrMode) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// BondingMode sets the mode for bonding.
+type BondingMode int
+
+const (
+	BondingUnset BondingMode = iota
+	BondingBalancedRR
+)
+
+// String implements fmt.Stringer.
+func (i BondingMode) String() string {
+	return [...]string{"", "balanced-rr"}[i]
+}
+
+// MarshalJSON implements json.Marshaler.
+func (i BondingMode) MarshalJSON() ([]byte, error) {
+	if i != BondingUnset {
+		return json.Marshal(i.String())
+	}
+
+	return []byte(JSONNull), nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (i *BondingMode) UnmarshalJSON(data []byte) error {
+	if string(data) == JSONNull {
+		*i = BondingUnset
+	} else {
+		var str string
+		if err := json.Unmarshal(data, &str); err != nil {
+			return err
+		}
+
+		toID := map[string]BondingMode{
+			"":            BondingUnset,
+			"balanced-rr": BondingBalancedRR,
+		}
+		mode, ok := toID[str]
+		if !ok {
+			return &json.UnmarshalTypeError{
+				Value: fmt.Sprintf("string %q", str),
+				Type:  reflect.TypeOf(i),
+			}
+		}
+		*i = mode
+	}
+
+	return nil
+}
+
 // HostCfg groups host specific configuration.
 type HostCfg struct {
-	IPAddrMode       IPAddrMode        `json:"network_mode"`
-	HostIP           *netlink.Addr     `json:"host_ip"`
-	DefaultGateway   *net.IP           `json:"gateway"`
-	DNSServer        *net.IP           `json:"dns"`
-	NetworkInterface *net.HardwareAddr `json:"network_interface"`
-	ProvisioningURLs *[]*url.URL       `json:"provisioning_urls"`
-	ID               *string           `json:"identity"`
-	Auth             *string           `json:"authentication"`
-	Timestamp        *time.Time        `json:"timestamp"`
+	IPAddrMode        IPAddrMode        `json:"network_mode"`
+	HostIP            *netlink.Addr     `json:"host_ip"`
+	DefaultGateway    *net.IP           `json:"gateway"`
+	DNSServer         *net.IP           `json:"dns"`
+	NetworkInterface  *net.HardwareAddr `json:"network_interface"`
+	ProvisioningURLs  *[]*url.URL       `json:"provisioning_urls"`
+	ID                *string           `json:"identity"`
+	Auth              *string           `json:"authentication"`
+	Timestamp         *time.Time        `json:"timestamp"`
+	NetworkInterfaces *[]*string        `json:"network_interfaces"`
```

Comment:
Now we have `NetworkInterface *net.HardwareAddr` and `NetworkInterfaces *[]*string`, do we need both? Or can we rename one? Might look confusing from a user's perspective.

Reply:
We want to move the "*net.HardwareAddr" stuff to use interface names. Effectively NetwokInterface is ignored when bonding is used. Changing that thought would require more work so we opted for the easy route

Reply:
ok, got it.